utils/land-patterns.stanza

```diff
@@ -1453,3 +1473,234 @@ public defn two-pin-chip-landpattern (length:Toleranced,
+  grid-locs(nr, nc, pitch, pitch)
+
+; A "FullMatrix" BGA is one where no pads are depopulated and the pad
+; columns and rows are all equidistant.
+public defstruct FullMatrix <: BGAModifier
+
+; A "StageredMatrix" BGA is one where every other pad is depopulated,
+; alternating across rows. The distance between adjacent pads on subsequent
+; rows will be pitch / sqrt(2).
+public defstruct StaggeredMatrix <: BGAModifier
+
+; An "EquilaterTriangleMatrix" is like a staggered matrix, but where the 
+; pads are all equistance from eachother. This means for any three adjacent pads
+; with two in the same row or column, the centers of the pads form an equilateral 
+; triangle.
+public defstruct EquilateralTriangleMatrix <: BGAModifier
+
+; A "PerimeterMatrix" is like a full matrix, but with a region of pads in the
+; center depopulated.
+public defstruct PerimeterMatrix <: BGAModifier:
+  rows:Range, ; the range of rows to depopulate
+  cols:Range, ; the range of columns to depopulate
+
+; A "ThermallyEnhancedMatrix" is like a PerimeterMatrix, except with some
+; region in the center remaining populated.
+public defstruct ThermallyEnhancedMatrix <: BGAModifier:
+  perimeter-rows:Range, ; the range of rows to depopulate
+  perimeter-cols:Range, ; the range of columns to depopulate
+  interior-rows:Range, ; the range of rows in the center to remain populated
+  interior-cols:Range, ; the range of columns in the center to remain populated
+
+defmethod depopulated? (m:StaggeredMatrix, r:Int, c:Int): 
+  ((r + 1) % 2) == (c % 2)
+
+defmethod depopulated? (m:EquilateralTriangleMatrix, r:Int, c:Int): 
+  ((r + 1) % 2) == (c % 2)
+
+defmethod depopulated? (m:PerimeterMatrix, r:Int, c:Int):
+  contains?(rows(m), r) and contains?(cols(m), c)
+
+defmethod depopulated? (m:ThermallyEnhancedMatrix, r:Int, c:Int):
+  (contains?(perimeter-rows(m), r) and contains?(perimeter-cols(m), c)) and not
+  (contains?(interior-rows(m), r) and contains?(interior-cols(m), c))
+
+defmethod locs (m:EquilateralTriangleMatrix, nr:Int, nc:Int, pitch:Double):
+  val vertical   = sqrt(3.0) * 0.5 * pitch
+  val horizontal = pitch * 0.5
+  grid-locs(nr, nc, horizontal, vertical)
+
+public defn make-bga-landpattern (num-rows:Int, ; the number of rows of package
+                                  num-cols:Int, ; the number of columns of the package
+                                  pitch:Double, ; the pitch of the solder bumps
+                                  ball-diameter:Double, ; the nominal/typical diameter of the solder balls
+                                  pad-type:BGAPadType,  ; whether the pads will be "solder mask defined" or not
+                                  package-length:Toleranced, ; the length of the package
+                                  package-width:Toleranced,  ; the width of the package
+                                  modifier:BGAModifier|((Int, Int) -> True|False), ; the modifier that controls which pads are depopulated
```

Comment:
Instead of `BGAModifier|((Int, Int) -> True|False)`, just have it accept `BGAModifier`.

Callers can call:
```
make-bga-landpattern(10, 10, 3.3, ...,  CustomDepop(depop))
```
instead, where `CustomDepop` takes a closure and creates a `BGAModifier` out of it.

Reply:
Done